Snapshot building unification

PLukas2018 opened the pull request:
DiffKemp contains multiple commands for creating snapshots (`build`, `build-kernel`, `llvm-to-snapshot`), this PR tries to unify  `build` command with `build-kernel` command (as discussed in #287 ). The clang options and LLVM passes are taken from `build-kernel` command, which is taken as the expected default behaviour for `build`.

The `build` command will now by default:
- compile file/project using `-O1 -Xclang -disable-llvm-passes`,
- run selected simplification passes using `opt`.

Using `--clang-drop` does not change the behaviour mentioned above only drops certain flags which are in the Makefile.

Same as before this PR, `--clang_append` can be used to specify other flags which should be added. You can also use it to change the optimization level, but it will still use `-Xclang -disable-llvm-passes` and run the default simplification passes.

By using `--no-opt-override` you can turn off the default behaviour, then the `-O1 -Xclang -disable-llvm-passes` flags are not used and the default simplification passes are not run. For the optimisation level is then used the one which is specified in Makefile or you can change it by `--clang-append`.

viktormalik commented on diffkemp/llvm_ir/compiler.py:
```diff
@@ -0,0 +1,46 @@
+Syntax of this module must be python2 compatible because it is checked
+when building diffkemp.bulding.cc_wrapper with RPython.
+"""
+from diffkemp.utils import get_opt_command
```
Could we move `get_opt_command` to `compiler.py`? I'd like to minimize the amount of modules which must be Python2-compatible.

PLukas2018 replied:
- I could move `get_opt_command` to `compiler.py`, but it would be necessary to also move `get_llvm_version` to `compiler.py` because it is called from `get_opt_command`, then the `utils.py` syntax would not have to be python2 compatible.
- The second option which comes to my mind, is to move the `opt_llvm` function from `compiler.py` to a different file (eg. `diffkemp/llvm_ir/optimiser.py`), because it is not used from `cc_wrapper` (by `rpython`), only the `get_clang_default_options` is used from `cc_wrapper`, but rpython requires all functions in the file and called from them to have python2 syntax. This would minimise the functions which would have to have python2 syntax, but there would be two new modules/files (`compiler.py` and eg. `optimiser.py`) which would both contain only one function for now.

Let me know, which option would you prefer.

viktormalik replied:
Option 1 doesn't make much sense to me as `get_llvm_version` is used from other modules, too.

So, I'd either go with option 2 or with moving `get_opt_command` to `compiler.py` and moving `get_llvm_version` to a separate file (`llvm-version.py`?). I don't really have a strong preference here.

PLukas2018 replied:
I used option 2 (moved the `opt_llvm` to `optimiser.py` ), problem with the third option was that the `get_opt_command` is also used from multiple modules.

viktormalik commented on diffkemp/diffkemp.py:
```diff
@@ -58,6 +59,8 @@ def build_c_project(args):
         if args.clang_drop is not None
         else "",
         wrapper_env_vars["llvm_link"]: args.llvm_link,
+        wrapper_env_vars["no_opt_override"]: ("1" if args.no_opt_override
+                                              else "0")
```
Isn't it possible to pass a boolean to the env variable? So that we could do
```suggestion
        wrapper_env_vars["no_opt_override"]: not args.no_opt_override
```

PLukas2018 replied:
It is possible to pass a boolean, but I would still need to do something like this in `cc_wrapper` when loading the value:
```python3
if os.environ.get(wrapper_env_vars["no_opt_override"]) == "True":
        no_opt_override = True
    else:
        no_opt_override = False
```

because passing `False` will pass "False" and `bool("False") == True`

viktormalik replied:
I see. Ok, let's stick with `1` and `0`.

viktormalik commented on diffkemp/diffkemp.py:
```diff
@@ -113,6 +116,16 @@ def build_c_project(args):
+    if not args.no_opt_override:
+        # Run llvm passes on created LLVM IR files.
+        with open(db_filename, "r") as db_file:
+            for line in db_file:
+                if line.startswith("o:"):
```
Nit: this saves one level of indentation:
```suggestion
            for line in [l for l in db_file if l.startswith("o:")]:
```
You could even meld the next line inside but that would be probably less readable due to line breaks:
```suggestion
            for line in [l.split(":")[1].rstrip() for l in db_file if l.startswith("o:")]:
```

PLukas2018 replied:
I used the first suggestion, the second was too long for one line.

viktormalik reviewed (commented):
Nice! Just a couple of comments.

I also didn't run the code so I'd like to do that before merging.

Note: I don't think that we have tests that try to run `diffkemp build`, perhaps something worth adding to be able to test PRs like this one.

PLukas2018 commented:
> Note: I don't think that we have tests that try to run `diffkemp build`, perhaps something worth adding to be able to test PRs like this one.

Yes, I think that we do not have any tests for `build` command. I think that in one PR, I tried to write some tests for the command, I tested creating a snapshot from of a file/parts of kernel, but that did not get merged because I think we want to get rid of depending on kernel sources when testing...
So, I do not really know what the tests should look like, if the repo should contain a small make-based project (eg. Makefile + one c file) which would be used for testing of the `build` command?

FrNecas commented:
> I also didn't run the code so I'd like to do that before merging.

I've been using this PR for building snapshots for experimenting with SMT solving and so far it works nicely and I haven't encountered any issues as far as the functionality goes.

viktormalik commented:
> > Note: I don't think that we have tests that try to run `diffkemp build`, perhaps something worth adding to be able to test PRs like this one.
> 
> Yes, I think that we do not have any tests for `build` command. I think that in one PR, I tried to write some tests for the command, I tested creating a snapshot from of a file/parts of kernel, but that did not get merged because I think we want to get rid of depending on kernel sources when testing... So, I do not really know what the tests should look like, if the repo should contain a small make-based project (eg. Makefile + one c file) which would be used for testing of the `build` command?

Yeah, I imagined something of that sort - our own toy Makefile-based project for testing (in future, we could add CMake, Meson, etc.).

viktormalik commented:
> > I also didn't run the code so I'd like to do that before merging.
> 
> I've been using this PR for building snapshots for experimenting with SMT solving and so far it works nicely and I haven't encountered any issues as far as the functionality goes.

Thanks, this is very useful!

PLukas2018 commented:
> > > Note: I don't think that we have tests that try to run `diffkemp build`, perhaps something worth adding to be able to test PRs like this one.
> > 
> > 
> > Yes, I think that we do not have any tests for `build` command. I think that in one PR, I tried to write some tests for the command, I tested creating a snapshot from of a file/parts of kernel, but that did not get merged because I think we want to get rid of depending on kernel sources when testing... So, I do not really know what the tests should look like, if the repo should contain a small make-based project (eg. Makefile + one c file) which would be used for testing of the `build` command?
> 
> Yeah, I imagined something of that sort - our own toy Makefile-based project for testing (in future, we could add CMake, Meson, etc.).

I added the our toy Makefile-based project and tests, also I added testing of compilation of the `cc-wrapper` with RPython to CI.

viktormalik commented:
> I added the our toy Makefile-based project and tests, also I added testing of compilation of the `cc-wrapper` with RPython to CI.

Nice! IIUC, these two could be posted as separate PRs. Let's do that so that we can merge them straight away and use them to verify this PR.

Thanks!

PLukas2018 commented:
> Nice! IIUC, these two could be posted as separate PRs. Let's do that so that we can merge them straight away and use them to verify this PR.

I find a problem with running it with Nix, I will try to fix it and open another PR with just the tests.
Btw I included in the tests some parts specific to this PR, so I will add the specific things in this PR after the PR with the tests will be merged.

PLukas2018 commented:
> I find a problem with running it with Nix.

Just an update on the nix problem: I found out that in the nix development environment (`nix develop`) when running from one Python script second one as a subprocess (using eg. `check_call`), if the second Python script is importing diffkemp it fails on  `ModuleNotFoundError`.

This is the problem with `build` command right now, because `bin/diffkemp build` calls `make` which calls python script `cc_wrapper` in which is imported diffkemp, but the module is not found.

As far as I was able to find, the thing is that `setuptoolsShellHook` sets an environment variable `NIX_PYTHONPATH` which contains a path to a directory which contains `*.pth` file which contains the path to the directory with diffkemp. When running python, nix uses the `NIX_PYTHONPATH` variable and sets python to search for modules in all directories which are written in the `.pth` file (so the diffkemp can be imported), but then it unsets the `NIX_PYTHONPATH` which is the reason why subprocesses from python script cannot find diffkemp anymore. 

So far, I did not find  good way how to handle this,  one possible solution would be to add the path to diffkemp directory to the environment variable `PYTHONPATH` then it works fine, I think that the best place to update the variable would be in the `flake.nix` file, but did not find yet if there is way how to get the path to a directory in which is the `flake.nix` file located.

So if somebody understands this better, I would greatly appreciate any insights or suggestions on how to resolve this issue.

PLukas2018 commented:
I solved the nix problem by adding `export PYTHONPATH="$(pwd):$PYTHONPATH"` to `flake.nix` which is not best solution but so far I did not find better.
I opened new PR #324 which contains the tests for build command, after the PR is merged this PR should be then rebased on it.

FrNecas commented:
I've been playing around with the changes a bit more and trying to build `dietlibc-0.34` (https://www.fefe.de/dietlibc/ ) using this PR results in an error thrown by opt when optimizing `start.ll` file:

```
opt: bin-x86_64/start.ll:1:1: error: expected top-level entity
# 1 "x86_64/start.S"
^
```
The content of `start.ll` is:

```
# 1 "x86_64/start.S"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 381 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "x86_64/start.S" 2
# 1 "./dietfeatures.h" 1
# 2 "x86_64/start.S" 2

.text
.global _start
_start:



 popq %rdi
 movq %rsp,%rsi
 pushq %rdi

 leaq 8(%rsi,%rdi,8),%rdx






 movq %rdx, environ(%rip)
 .hidden environ
# 48 "x86_64/start.S"
 call stackgap
 .hidden stackgap
# 58 "x86_64/start.S"
 movq %rax, %rdi
# 70 "x86_64/start.S"
 call exit

 hlt
.Lstart:
 .size _start,.Lstart-_start
 .section .note.GNU-stack,"",@progbits
```

This works correctly on master because opt is not run there. It seems like that opt requires e.g. a function on the highest level which is not the case in this example (I am not even sure where this LLVM IR is coming from tbh). Maybe in such cases, diffkemp should catch the opt error and leave the file unoptimized?

lenticularis39 commented:
> This works correctly on master because opt is not run there. It seems like that opt requires e.g. a function on the highest level which is not the case in this example (I am not even sure where this LLVM IR is coming from tbh). Maybe in such cases, diffkemp should catch the opt error and leave the file unoptimized?

This seems to be x86 assembly rather than LLVM IR, not sure why that ended up there. Can you post the original command that generated the file?

PLukas2018 commented:
> This works correctly on master because opt is not run there. It seems like that opt requires e.g. a function on the highest level which is not the case in this example (I am not even sure where this LLVM IR is coming from tbh). Maybe in such cases, diffkemp should catch the opt error and leave the file unoptimized?

Thank you for the report. I will add the catching of error, I think it is good idea to do that if the optimization for some reason was unsuccessful rather than to end the build command.

PLukas2018 commented:
> This seems to be x86 assembly rather than LLVM IR, not sure why that ended up there. Can you post the original command that generated the file?

The `cc_wrapper` was called with 
```bash
-I. -isystem include -Os -fstrict-aliasing -momit-leaf-frame-pointer -mfancy-math-387  -W -Wall -Wextra -Wchar-subscripts -Wmissing-prototypes -Wmissing-declarations -Wno-switch -Wno-unused -Wredundant-decls -Wshadow -fpie  -c x86_64/start.S -Wa,--noexecstack -fno-pie -o bin-x86_64/start.o
```
cc_wrapper compiles it like this:
```bash
clang -I. -isystem include -Os -fstrict-aliasing -momit-leaf-frame-pointer -mfancy-math-387 -W -Wall -Wextra -Wchar-subscripts -Wmissing-prototypes -Wmissing-declarations -Wno-switch -Wno-unused -Wredundant-decls -Wshadow -fpie -c x86_64/start.S -Wa,--noexecstack -fno-pie -o bin-x86_64/start.ll -S -emit-llvm -g -fdebug-macro -Wno-format-security -O1 -Xclang -disable-llvm-passes
```
The compilation is successful even tho the output is not LLVM IR (but `-emit-llvm` is used in flags).
In this PR I added running opt with certain passes on files which `cc_wrapper` adds to `diffkemp-wdb` (compiles to LLVM IR), and because this is added to `diffkemp-wdb` I run on it `opt` which results in the error. 

I am not sure how `*.S` files should be handled, should they even be compiled by `cc_wrapper`?

lenticularis39 commented:
> I am not sure how `*.S` files should be handled, should they even be compiled by `cc_wrapper`?
>

Assembly files should ideally be omitted by the wrapper, since it is not possible to get any LLVM IR from them. IIRC the current version of cc_wrapper does not account for assembly source, only for object files and source files.

PLukas2018 commented:
> Assembly files should ideally be omitted by the wrapper, since it is not possible to get any LLVM IR from them. IIRC the current version of cc_wrapper does not account for assembly source, only for object files and source files.

Currently, there is not check what the source file is (C, assembly), I opened PR #326 which checks if the source file is C source file. This should solve the error, but I would still rather leave the catch for running `opt` tool if in other library/project would be some other problems.

viktormalik commented on diffkemp/llvm_ir/single_c_builder.py:
```diff
@@ -13,29 +15,41 @@ class SingleCBuilder(SingleLlvmFinder):
-        command = [self.clang, "-S", "-emit-llvm", "-g", self.c_file_name,
-                   "-o", self.llvm_file_name]
+        command = [self.clang, self.c_file_name, "-o", self.llvm_file_name]
+        # Note: extending with the default options must be before
+        # extending with the clang_append option
+        # so the optimisations flags can be redefined
+        # (it looks like clang uses last specified optimisaton flag).
```
What about this:
```suggestion
        # Note: clang uses the last specified optimization level so
        # extending with the default options must be done before
        # extending with the clang_append option.
```
(probably better wrapped)

viktormalik reviewed (commented):
Looks good to me, just a couple of minor comments. Thanks!

PLukas2018 commented on diffkemp/building/CMakeLists.txt:
```diff
@@ -20,9 +20,10 @@ if (NOT DEFINED ENV{WITHOUT_RPYTHON})
     add_custom_command(OUTPUT ${CC_WRAPPER}
                        DEPENDS ${CC_WRAPPER_SOURCE}
-                       COMMAND ${RPYTHON} ${CC_WRAPPER_SOURCE} "--output=${CC_WRAPPER}"
+                       COMMAND PYTHONPATH=${CMAKE_SOURCE_DIR} ${RPYTHON} ${CC_WRAPPER_SOURCE} "--output=${CC_WRAPPER}"
```
I added the `$PYTHONPATH` to the end, so we do not have to think about this if we decide to use external modules in the future.

PLukas2018 commented on flake.nix:
```diff
@@ -120,6 +120,13 @@
             shellHook = ''
               export -f setuptoolsShellHook runHook _eval _callImplicitHook
+
+              # Adding current (diffkemp) directory to PYTHONPATH,
+              # the `diffkemp build` subcommand does not work without it
+              # - `cc_wrapper.py` ends with `ModuleNotFoundError` because
+              # `setuptoolsShellHook` does not make diffkemp package
+              # importable for subprocesses called from python.
+              export PYTHONPATH="$(pwd):$PYTHONPATH"
```
Yes, it probably won't work when run from a different directory (not sure if it si possible), but so far I did not think up a better approach.

PLukas2018 commented:
I addressed the comments.

Btw the [CI failed](https://github.com/diffkemp/diffkemp/actions/runs/8648498687/job/23712463316), which happened to me already in the past, I am not sure if it is something connected with this PR, I think it should not be, because when I reran the CI last time it was ok, I will rerun it again this time too.
It fails on missing kernels in regression tests, but on closer look, there is an error when the cached kernels are being restored. I will probably look closer at this problem (Could it be because we have not yet updated actions to a newer one which uses Node.js 20, probably not, ...).

viktormalik reviewed (approved):
Nice, thanks for this work!
